lib/msal-common/src/utils/ProtocolUtils.ts

```diff
@@ -15,23 +28,52 @@ export class ProtocolUtils {
      * @param userState 
      * @param randomGuid 
      */
-    static setRequestState(userState: string, randomGuid: string): string {
-        return !StringUtils.isEmpty(userState) ? `${randomGuid}${Constants.RESOURCE_DELIM}${userState}` : randomGuid;
+    static setRequestState(userState: string, randomGuid: string, cryptoObj: ICrypto): string {
+        const libraryState = ProtocolUtils.generateLibraryState(randomGuid, cryptoObj);
+        return !StringUtils.isEmpty(userState) ? `${libraryState}${Constants.RESOURCE_DELIM}${userState}` : libraryState;
     }
 
     /**
-     *
-     * Extracts user state value from the state sent with the authentication request.
-     * @returns {string} scope.
-     * @ignore
+     * Generates the state value used by the library.
+     * @param randomGuid 
+     * @param cryptoObj 
+     * 
+     * @returns Base64 encoded string representing the state
      */
-    static getUserRequestState(serverResponseState: string): string {
-        if (!StringUtils.isEmpty(serverResponseState)) {
-            const splitIndex = serverResponseState.indexOf(Constants.RESOURCE_DELIM);
-            if (splitIndex > -1 && splitIndex + 1 < serverResponseState.length) {
-                return serverResponseState.substring(splitIndex + 1);
-            }
+    static generateLibraryState(randomGuid: string, cryptoObj: ICrypto): string {
+        const stateObject: LibraryStateObject = {
+            id: randomGuid,
+            ts: TimeUtils.nowSeconds()
+        };
+
+        const stateString = JSON.stringify(stateObject);
+
+        return cryptoObj.base64Encode(stateString);
+    }
+
+    /**
+     * Extracts user state value and library state value 
+     * from the state sent with the authentication request.
+     * @param state 
+     * @param cryptoObj 
+     */
+    static parseRequestState(state: string, cryptoObj: ICrypto): RequestStateObject {
+        if (StringUtils.isEmpty(state)) {
+            throw ClientAuthError.createInvalidStateError(state);
+        }
+
+        try {
+            const splitState = decodeURIComponent(state).split(Constants.RESOURCE_DELIM);
+            const libraryState = splitState.splice(0, 1)[0];
```

Comment:
Why do we need to splice here?

Reply:
Comment might be helpful here @pkanher617

Reply:
Will add a comment here. @tnorling This is so that we can do the `.join()` on the next line and preserve the user state.

Reply:
Would it not be cleaner to use `splitState[0]` here and then do `splitState.slice(1).join(Constants.RESOURCE_DELIM)` below instead of mutating the original array?

Reply:
done.